### PR TITLE
[ICONS] let users set custom title to icons

### PR DESCRIPTION
**Pull request by Shijir**
Tested on: Chrome, FF, Safari, Edge, IE 11

Signed-off-by: Shijir Tsogoo <stsogoo@vmware.com>

### CODE REVIEW

File: src/clarity-icons/clarity-icons-element.ts [hunk 1]
```diff
@@ -20,28 +22,24 @@ export function ClarityIconElement() {
...
-        }
+const replaceTitleWithCustom = function(currentShapeTemplate: string, currentTitleAttrVal: string,
+                                        clrIconUniqId: string) {
+    // if the title tag doesn't exist right after that first svg tag, we will inject a
+    if (currentShapeTemplate.indexOf("<title") > -1 && currentShapeTemplate.indexOf("</title>") > -1) {
+        // if title tag exists, remove that fist
+        currentShapeTemplate = currentShapeTemplate.replace(/<title[^>]*>([^<]+)<\/title>/i, "");
     }
+
+    const openingSVGTag = currentShapeTemplate.match(/<svg\b[^>]*>/i)[0];  // will find the first occurrence of svg tag
```
Comment (mathisscott) [inline src/clarity-icons/clarity-icons-element.ts]: Icons can be added that are not SVG. Does this code still support that?

Comment (Shijir) [reply src/clarity-icons/clarity-icons-element.ts]: oh totally forgot about that. I will fix that.

File: src/clarity-icons/clarity-icons-element.ts [hunk 2]
```diff
@@ -20,28 +22,30 @@ export function ClarityIconElement() {
...
-        if (ClarityIconsApi.instance.has(shape)) {
-            element.innerHTML = ClarityIconsApi.instance.get(shape);
-        } else {
-            console.error(`'${shape}' is not found in the Clarity Icons set.`);
-            element.innerHTML = ClarityIconsApi.instance.get("error");
-        }
+    // if the title tag doesn't exist right after that first svg tag, we will inject a
+    if (currentShapeTemplate.indexOf("<title") > -1 && currentShapeTemplate.indexOf("</title>") > -1) {
+        // if title tag exists, remove that fist
+        currentShapeTemplate = currentShapeTemplate.replace(/<title[^>]*>([^<]+)<\/title>/i, "");
```
Comment (youdz) [inline src/clarity-icons/clarity-icons-element.ts]: Isn't it risky to nuke the title instead of replacing its content directly?
I could see it being an issue on custom SVGs the consumer adds to Clarity, they might have extra attributes on the title that we are not preserving here.

Comment (youdz) [inline src/clarity-icons/clarity-icons-element.ts]: Another comment here: you literally just grabbed the indexes of the beginning and end of the title tag, do you really have to do another regex search on a potentially large string?
```typescript
currentShapeTemplate = currentShapeTemplate.slice(0, titleStart) + currentShapeTemplate.slice(titleEnd+8)
```
I believe that would be more performant, and it's pretty free.

File: src/clarity-icons/clarity-icons-element.ts [hunk 3]
```diff
@@ -20,28 +22,30 @@ export function ClarityIconElement() {
...
 ClarityIconElement.prototype = Object.create(
     HTMLElement.prototype, {constructor: {configurable: true, writable: true, value: ClarityIconElement}});
 
 ClarityIconElement.prototype.constructor = ClarityIconElement;
 
-const generateIcon = function(element: any, shape: string) {
-    shape = shape.split(/\s/)[0];
+const replaceTitleWithCustom = function(currentShapeTemplate: string, currentTitleAttrVal: string,
+                                        clrIconUniqId: string) {
+    const hasSVGtag = currentShapeTemplate.match(/<svg\b[^>]*>/i);  // will find the first occurrence of svg tag
```
Comment (youdz) [inline src/clarity-icons/clarity-icons-element.ts]: Do you need a regex here? `indexOf()` with plain strings is way faster than regex matching, and I think matching `"<svg"` is enough, isn't it?

File: src/clarity-icons/clarity-icons-element.ts [hunk 4]
```diff
@@ -20,28 +22,30 @@ export function ClarityIconElement() {
...
-            element.innerHTML = ClarityIconsApi.instance.get("error");
-        }
+    // if the title tag doesn't exist right after that first svg tag, we will inject a
+    if (currentShapeTemplate.indexOf("<title") > -1 && currentShapeTemplate.indexOf("</title>") > -1) {
+        // if title tag exists, remove that fist
+        currentShapeTemplate = currentShapeTemplate.replace(/<title[^>]*>([^<]+)<\/title>/i, "");
     }
+
+    const openingSVGTag = hasSVGtag[0];
+    const remainingSVG = currentShapeTemplate.split(/<svg\b[^>]*>/i)[1];
```
Comment (youdz) [inline src/clarity-icons/clarity-icons-element.ts]: You literally did this exact regex match 15 lines above. You should just use that result.

File: src/clarity-icons/clarity-icons-element.ts [hunk 5]
```diff
@@ -55,25 +59,99 @@ const setIconSize = function(element: any, size: string) {
...
+            return;
+        }
+
+        this.currentShapeAttrVal = shapeAttrValue;
+
+        if (ClarityIconsApi.instance.has(this.currentShapeAttrVal)) {
+            this.currentShapeTemplate = ClarityIconsApi.instance.get(this.currentShapeAttrVal);
+        } else {
+            console.error(`'${this.currentShapeAttrVal}' is not found in the Clarity Icons set.`);
+            host.innerHTML = ClarityIconsApi.instance.get("error");
```
Comment (youdz) [inline src/clarity-icons/clarity-icons-element.ts]: I believe you still want to put the user-given title even if you can't find the icon. That's super easy for us to do and helps mitigate the error when users can hover over the error icon to see what it was supposed to mean. Screen readers will even have the normal experience.

File: src/clarity-icons/clarity-icons-element.ts [hunk 6]
```diff
@@ -55,25 +59,99 @@ const setIconSize = function(element: any, size: string) {
...
+    // before this connectedCallback method called only once when the custom element is inserted into the DOM.
+    // So we could check whether the attribute values really changed or not.
+    // If not, we don't need to execute the same codes again.
     const host = this as HTMLElement;
 
     if (host.hasAttribute("shape")) {
-        generateIcon(host, host.getAttribute("shape"));
+        const shapeAttrValue = host.getAttribute("shape").split(/\s/)[0];
+
+        if (shapeAttrValue === this.currentShapeAttrVal) {
```
Comment (youdz) [inline src/clarity-icons/clarity-icons-element.ts]: You inlined this from the function, but this test makes no sense for the `connectedCallback`, `this.currentShapeAttrVal` will always be undefined. It's only here for when the shape attribute changes.

File: src/clarity-icons/clarity-icons-element.ts [hunk 7]
```diff
@@ -55,25 +59,99 @@ const setIconSize = function(element: any, size: string) {
...
+    if (host.hasAttribute("title")) {
+        const titleAttrValue = host.getAttribute("title");
+
+        if (this.currentTitleAttrVal !== titleAttrValue) {
+            this.currentTitleAttrVal = titleAttrValue;
+        }
+
+        if (!this.currentShapeAttrVal) {
+            return;
+        }
```
Comment (youdz) [inline src/clarity-icons/clarity-icons-element.ts]: Why is this here, in the middle of the title code? If the icon has a title but no shape, you're never setting its size, but if the title has no title and no shape you do set the size correctly?

File: src/clarity-icons/clarity-icons-element.ts [hunk 8]
```diff
@@ -55,25 +59,99 @@ const setIconSize = function(element: any, size: string) {
...
+
+    this.injectTemplate(host);
 };
 
 ClarityIconElement.prototype.attributeChangedCallback = function(attributeName: string, oldValue: string,
                                                                  newValue: string) {
     const host = this as HTMLElement;
 
+    if (!this.initialized) {
+        this.clrIconUniqId = "_clr_icon_" + clrIconId;
```
Comment (youdz) [inline src/clarity-icons/clarity-icons-element.ts]: Why is this set in the `attributeChangedCallback`? This is initialization, it clearly belongs to `connectedCallback`. You wouldn't even need the `this.initialized` property and test.

File: src/clarity-icons/clarity-icons-element.ts [hunk 9]
```diff
@@ -55,25 +59,99 @@ const setIconSize = function(element: any, size: string) {
...
+            host.innerHTML = ClarityIconsApi.instance.get("error");
+            return;
+        }
+    }
+    if (attributeName === "title") {
+        this.currentTitleAttrVal = newValue;
+
+        if (!this.currentShapeAttrVal) {
+            return;
+        }
```
Comment (youdz) [inline src/clarity-icons/clarity-icons-element.ts]: That return I understand slightly more, but even then I don't see the point. And why is this only for the title and not the size? They should have exactly the same problems with no shape.

Just fallback "gracefully" to the error icon with the correct title and size, it's the most friendly behavior and we can have it for free.

File: src/clarity-icons/clarity-icons-element.ts [hunk 10]
```diff
@@ -55,25 +59,99 @@ const setIconSize = function(element: any, size: string) {
...
-        generateIcon(host, newValue);
+        this.currentShapeAttrVal = newValue.split(/\s/)[0];
+
+        if (ClarityIconsApi.instance.has(this.currentShapeAttrVal)) {
+            this.currentShapeTemplate = ClarityIconsApi.instance.get(this.currentShapeAttrVal);
+        } else {
+            console.error(`'${this.currentShapeAttrVal}' is not found in the Clarity Icons set.`);
+            host.innerHTML = ClarityIconsApi.instance.get("error");
+            return;
+        }
```
Comment (youdz) [inline src/clarity-icons/clarity-icons-element.ts]: Why did you inline this but kept the external function for the icon size? You now have this code duplicated in two methods of `ClarityIconElement`, and I don't understand why you had to do it.

File: src/clarity-icons/clarity-icons-element.ts [hunk 11]
```diff
@@ -55,25 +59,99 @@ const setIconSize = function(element: any, size: string) {
...
     if (attributeName === "size") {
         setIconSize(host, newValue);
     }
+
+    this.injectTemplate(host);
+};
+
+ClarityIconElement.prototype.injectTemplate = function(hostElement: any) {
+    if (this.currentTitleAttrVal) {
+        hostElement.setAttribute("aria-labelledby", this.clrIconUniqId);
```
Comment (youdz) [inline src/clarity-icons/clarity-icons-element.ts]: When called from the `connectedCallback`, it sets `aria-labelledby` to undefined because the id is not ready yet. See my comment on the initialization happening in the wrong place.

File: src/clarity-icons/clarity-icons.spec.ts [hunk 12]
```diff
@@ -373,15 +356,134 @@ describe("ClarityIcons", () => {
...
+
+            const clarityIcon = document.createElement("clr-icon");
+            clarityIcon.setAttribute("shape", "test-shape");
+            clarityIcon.setAttribute("title", "my-custom-title");
+
+            expect(removeWhitespace(clarityIcon.innerHTML)).toBe(removeWhitespace(testShape));
+        });
+
+        it("should not inject anything if the custom title is not given", () => {
+            const customTitle = "my-custom-title";
```
Comment (youdz) [inline src/clarity-icons/clarity-icons.spec.ts]: You're not using `customTitle` here. Happens to all of us with test copy-paste. 😛

File: src/clarity-icons/clarity-icons.spec.ts [hunk 13]
```diff
@@ -373,15 +356,134 @@ describe("ClarityIcons", () => {
...
+
+            const clarityIcon = document.createElement("clr-icon");
+            clarityIcon.setAttribute("shape", "test-shape");
+
+            expect(clarityIcon.hasAttribute("aria-labelledby")).toBeFalsy();
+            expect(removeWhitespace(clarityIcon.innerHTML)).toBe(removeWhitespace(testShape));
+        });
+
+        it("should inject a custom title into the template if title attribute is present", () => {
+            const timeStart = performance.now();
```
Comment (youdz) [inline src/clarity-icons/clarity-icons.spec.ts]: Leftover performance helper.

File: src/clarity-icons/clarity-icons.spec.ts [hunk 14]
```diff
@@ -373,15 +356,134 @@ describe("ClarityIcons", () => {
...
+
+            clarityIcon.setAttribute("shape", "angle");
+            const timeAngleShape = performance.now();
+            console.log("Icon load time: " + (timeAngleShape - timeStart));
+
+            clarityIcon.setAttribute("title", customTitle);
+            const timeCustomTitle = performance.now();
+            console.log("Icon load time: " + (timeCustomTitle - timeStart));
+
+            // tslint:disable-next-line:no-string-literal
```
Comment (youdz) [inline src/clarity-icons/clarity-icons.spec.ts]: If you want to avoid this, just declare a `ClarityIconElement` interface that extends `HTMLElement` with our few custom properties. It'll be 100 times cleaner and you can use it in the actual code too, allowing you to get proper type checking.

File: src/clarity-icons/clarity-icons.spec.ts [hunk 15]
```diff
@@ -373,15 +356,134 @@ describe("ClarityIcons", () => {
...
+            const clrIconUniqId = (clarityIcon as any)["clrIconUniqId"];
+            const testShape = giveAngleShapeTitle(clrIconUniqId, customTitle);
+
+            expect(removeWhitespace(clarityIcon.innerHTML)).toBe(removeWhitespace(testShape));
+        });
+
+        it("should inject custom title if given and template doesn't contain title tag", () => {
```
Comment (youdz) [inline src/clarity-icons/clarity-icons.spec.ts]: I think you're missing a test to check that if there is a title in the SVG, it is properly *replaced* by the one given as attribute. You don't want 2 title tags in your SVG.

File: src/clarity-icons/clarity-icons-element.ts (hunk 3 above)
Comment (Shijir) [reply src/clarity-icons/clarity-icons-element.ts]: @youdz I've updated replaceTitleWithCustom function. This updated version doesn't rely on the regex; it doesn't remove elements such as title from user's template. It converts the template into a dom element and makes all the required changes into it before injecting it into the icon.

File: src/clarity-icons/clarity-icons-element.ts (hunk 2 above)
Comment (Shijir) [reply src/clarity-icons/clarity-icons-element.ts]: I have addressed it in the updated version of the method.

File: src/clarity-icons/clarity-icons-element.ts (hunk 6 above)
Comment (Shijir) [reply src/clarity-icons/clarity-icons-element.ts]: We set value to `this.currentShapeAttrVal` in `attributeChangedCallback()`, which gets invoked before `connectedCallback()`. So it won't always be undefined.

File: src/clarity-icons/clarity-icons-element.ts (hunk 5 above)
Comment (Shijir) [reply src/clarity-icons/clarity-icons-element.ts]: Good point! I will implement that!

File: src/clarity-icons/clarity-icons-element.ts (hunk 8 above)
Comment (Shijir) [reply src/clarity-icons/clarity-icons-element.ts]: What I figured out was that the `attributeChangedCallback()` is invoked before `connectedCallback()`. So I had to initialize the properties here with `this.initialized` guard. The `connectedCallback()` method gets called only once when it's actually inserted into the document tree. Of course, if any of the attributes are changed after the element is inserted into the dom, `attributeChangedCallback()` will be called after `connectedCallback()` as well.

File: src/clarity-icons/clarity-icons-element.ts (hunk 10 above)
Comment (Shijir) [reply src/clarity-icons/clarity-icons-element.ts]: I have now `injectErrorTemplate()` function takes care of this part of code.

File: src/clarity-icons/clarity-icons.spec.ts (hunk 13 above)
Comment (Shijir) [reply src/clarity-icons/clarity-icons.spec.ts]: I intentionally left these logs to see how long it's taking to load an icon. 😛

File: src/clarity-icons/clarity-icons.spec.ts (hunk 15 above)
Comment (Shijir) [reply src/clarity-icons/clarity-icons.spec.ts]: Thanks! I have added in a test with a following description: `it("should update existing title if custom title is specified")`

File: src/clarity-icons/clarity-icons.spec.ts (hunk 12 above)
Comment (Shijir) [reply src/clarity-icons/clarity-icons.spec.ts]: fixed that.

File: src/clarity-icons/clarity-icons-element.ts (hunk 8 above)
Comment (Shijir) [reply src/clarity-icons/clarity-icons-element.ts]: Ideally, I needed to initialize them in the constructor. But `this` in the constructor and the ones in any of the prototype methods of `ClarityIconElement` were pointing to different object. But I figured out why that was so and made changes accordingly to fix that issue.

File: src/clarity-icons/clarity-icons-element.ts (hunk 9 above)
Comment (Shijir) [reply src/clarity-icons/clarity-icons-element.ts]: I have moved the size testing part to the area above both the shape attribute check. So we get hold of the size attribute before shape and title attributes.

As to falling back to the error icon, in cases when title attribute value happens to change right before shape attribute changes, why do we need to inject the error icon?

Comment (youdz) [reply src/clarity-icons/clarity-icons-element.ts]: The order doesn't really matter, what I'm saying right now is that if I put a correct title but a wrong shape, I get an error icon _without_ my title. You should at least display the error icon with the correct title.
You noted this was a good idea and said you would do it, but I still don't see it anywhere.

File: src/clarity-icons/clarity-icons-element.ts [hunk 16]
```diff
@@ -18,30 +20,45 @@ if (typeof Reflect === "object") {
...
-        }
+    if (!svgElement) {
+        return currentShapeTemplate;
     }
+
+    let titleElement = svgElement.querySelector("title");
+
+    if (titleElement) {
+        titleElement.setAttribute("id", clrIconUniqId);
+        titleElement.innerHTML = currentTitleAttrVal;
```
Comment (youdz) [inline src/clarity-icons/clarity-icons-element.ts]: I think you want `textContent` here, not `innerHTML`. It's safer and you're not supporting HTML in SVG titles anyway.

File: src/clarity-icons/clarity-icons-element.ts [hunk 17]
```diff
@@ -18,30 +20,45 @@ if (typeof Reflect === "object") {
...
+    } else {
+        titleElement = document.createElement("title");
+        titleElement.setAttribute("id", clrIconUniqId);
+        titleElement.innerHTML = currentTitleAttrVal;
```
Comment (youdz) [inline src/clarity-icons/clarity-icons-element.ts]: Same here.

File: src/clarity-icons/clarity-icons-element.ts [hunk 18]
```diff
@@ -31,21 +37,28 @@ ClarityIconElement.prototype.constructor = ClarityIconElement;
...
+        svgElement.insertAdjacentElement("afterbegin", titleElement);
     }
 
-    const openingSVGTag = hasSVGtag[0];
-    const remainingSVG = currentShapeTemplate.split(/<svg\b[^>]*>/i)[1];
-    return openingSVGTag + `<title id="${clrIconUniqId}">${currentTitleAttrVal}</title>` + remainingSVG;
+    return tempContainer.innerHTML;
```
[the remark this answers is not in the document]
Comment (mathisscott) [reply src/clarity-icons/clarity-icons-element.ts]: We should not be going from `String -> Html -> String -> Html`

I think we can do what we want with basic (and _fast_!) string operations and then apply it to the DOM once.

File: src/clarity-icons/clarity-icons-element.ts (hunk 6 above)
Comment (Shijir) [reply src/clarity-icons/clarity-icons-element.ts]: I do think there are cases that users may want to manipulate `clr-icon` elements before or without injecting them into the DOM tree. In those cases, `clr-icon` element's attributes should update properly. Also, in environments with simulated or fake DOM tree, clr-icon attributes probably wouldn't work. For example, if I do:
```
let myIcon = document.createElement("clr-icon");
myIcon.setAttribute("shape", "user");
console.log(myIcon.innerHTML); // no SVG template as we have to now append to the DOM.
```

File: src/clarity-icons/clarity-icons-element.ts (hunk 9 above)
Comment (Shijir) [reply src/clarity-icons/clarity-icons-element.ts]: Wait 🤔  I've actually addressed that issue. I have even added a test for the case:
```
it("should inject error shape with title if icon is not found", ...);
```

File: src/clarity-icons/clarity-icons-element.ts (hunk 7 above)
Comment (Shijir) [reply src/clarity-icons/clarity-icons-element.ts]: This is here to escape calling `this.injectTemplate(host)` as a user hasn't yet provided the shape at this point.

File: src/clarity-icons/clarity-icons-element.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (Shijir) [reply src/clarity-icons/clarity-icons-element.ts]: @youdz @mathisscott I resorted to DOM operations as I thought this would be more robust and reliable than using regexes and string operations. When I do benchmark tests, this approach happened to show slightly better performance than the previous string manipulation approach I had. I think it doesn't perform badly because we are doing all the DOM operations **before** injecting it into the actual DOM tree. So there is no deep DOM traversal or no DOM reflow activity.  Also, in terms of readability, it looks much cleaner and easier to read IMO.

The only hesitation or question over this approach I have now is: can we accomplish better results if we find better string manipulating operations? I think better string manipulating operations means we should avoid regexes as much as possible because as soon as I use them, the benchmark tests are not showing better results, but actually worse. And manipulating string HTML template without regex to reach the desired result can be very difficult.

[the remark this answers is not in the document]
Comment (Shijir) [reply src/clarity-icons/clarity-icons-element.ts]: Or we could also end up achieving better results if we use both dom operations and string manipulating operators. That's what I've been pondering, testing, and experimenting.

[the remark this answers is not in the document]
Comment (mathisscott) [reply src/clarity-icons/clarity-icons-element.ts]: I would say:

• avoid regex for this

**Comment by mathisscott** [comment]
@Shijir 
What is the status of this PR?

**Comment by Shijir** [comment]
@mathisscott 
It's ready for reviews. I've just requested reviews from other team members. Also, just talked to @youdz about this issue. He said he would go over it as he is catching up on pull requests.

File: src/clarity-icons/clarity-icons.scss [hunk 19]
```diff
@@ -95,11 +95,25 @@ clr-icon {
...
         width: 100%;
         display: block;
         transition: inherit;
+        title {
+            display: none;
+        }
+    }
+    & > span[id^=_clr_icon_] {
+        position: absolute;
+        left: -10000px;
```
Comment (adityarb88) [inline src/clarity-icons/clarity-icons.scss]: This will cause scrollbars to appear if the display is changed from ltr to rtl. Could you reuse this class instead? https://github.com/vmware/clarity/blob/master/src/clarity-angular/popover/common/_popover.clarity.scss#L6

File: src/clarity-icons/clarity-icons.scss [hunk 20]
```diff
@@ -198,7 +198,6 @@ clr-icon {
         top: -1px !important;
 
         overflow: hidden !important;
-        visibility: hidden !important;
```
Comment (adityarb88) [inline src/clarity-icons/clarity-icons.scss]: Why are we removing `visibility: hidden`?

Comment (Shijir) [reply src/clarity-icons/clarity-icons.scss]: The main reason was that `visibility: hidden;` makes screen readers ignore the content text. 
https://webaim.org/techniques/css/invisiblecontent/

Comment (adityarb88) [reply src/clarity-icons/clarity-icons.scss]: Yep. Totally forgot about that. This is different than popovers.
